server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java

```diff
@@ -0,0 +1,350 @@
+
+        if (oldBytes == null) {
+          created = true;
+        } else if (!Arrays.equals(oldBytes, bytes)) {
+          throw new IAE("Cannot reannounce different values under the same path");
+        }
+      }
+    }
+
+    if (created) {
+      try {
+        createAnnouncement(path, bytes);
+      }
+      catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  public void update(final String path, final byte[] bytes)
+  {
+    synchronized (toAnnounce) {
+      if (!started) {
+        // removeParentsIfCreated is not relevant for updates; use dummy value "false".
+        toUpdate.add(new Announceable(path, bytes, false));
+        return;
+      }
+    }
+
+    byte[] oldBytes = announcedPaths.get(path);
+
+    if (oldBytes == null) {
+      throw new ISE("Cannot update a path[%s] that hasn't been announced!", path);
+    }
+
+    synchronized (toAnnounce) {
+      try {
+        if (!Arrays.equals(oldBytes, bytes)) {
+          announcedPaths.put(path, bytes);
+          updateAnnouncement(path, bytes);
+        }
+      }
+      catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private void createAnnouncement(final String path, byte[] value) throws Exception
+  {
+    curator.create().compressed().withMode(CreateMode.EPHEMERAL).inBackground().forPath(path, value);
+  }
+
+  private void updateAnnouncement(final String path, final byte[] value) throws Exception
+  {
+    curator.setData().compressed().inBackground().forPath(path, value);
+  }
+
+  /**
+   * Unannounces an announcement created at path.  Note that if all announcements get removed, the Announcer
+   * will continue to have ZK watches on paths because clearing them out is a source of ugly race conditions.
+   * <p/>
+   * If you need to completely clear all the state of what is being watched and announced, stop() the Announcer.
+   *
+   * @param path the path to unannounce
+   */
+  public void unannounce(String path)
+  {
+    log.info("unannouncing [%s]", path);
+    final byte[] value = announcedPaths.remove(path);
+
+    if (value == null) {
+      log.error("Path[%s] not announced, cannot unannounce.", path);
+      return;
+    }
+
+    try {
+      curator.transaction().forOperations(curator.transactionOp().delete().forPath(path));
+    }
+    catch (KeeperException.NoNodeException e) {
+      log.info("node[%s] didn't exist anyway...", path);
+    }
+    catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void startCache(NodeCache cache)
+  {
+    try {
+      cache.start();
+    }
+    catch (Exception e) {
+      CloseQuietly.close(cache);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void createPath(String parentPath, boolean removeParentsIfCreated)
+  {
+    try {
+      curator.create().creatingParentsIfNeeded().forPath(parentPath);
+      if (removeParentsIfCreated) {
+        pathsCreatedInThisAnnouncer.add(parentPath);
```

Comment:
The path is added `pathsCreatedInThisAnnouncer` regardless of whether it was actually created two lines above or already existed?

Reply:
I think so, what's your concern?

Reply:
Because then the path is not actually `paths**Created**InThisAnnouncer`

Reply:
This should be resolved before merge